File: problem_solving_algo_and_ds/ruby/data_structures/par_checker.rb
require_relative 'stack'

def par_checker(symbol_string)
  if symbol_string.length.odd?
    return false
  end

  stack = Stack.new

  symbol_string.chars.each do |char|
    if char == "("
      stack.push(char)
    else
      if !stack.is_empty
        stack.pop
      end
    end
  end

  if stack.is_empty
    return true
  else
    return false
  end
end

puts par_checker('((()))')
puts par_checker('(()')

File: problem_solving_algo_and_ds/python/data_structures/par_checker.py
from stack import *


def par_checker(symbol_string):
    stack = Stack()

    for paren in symbol_string:
        if paren == "(":
            stack.push(paren)
        else:
            if not stack.isEmpty():
                stack.pop()

    return stack.isEmpty()


print(par_checker('((()))'))
print(par_checker('(()'))
